stdlib/public/core/KeyPath.swift

```diff
@@ -3713,6 +3878,25 @@ internal func _instantiateKeyPathBuffer(
     endOfReferencePrefixComponent.storeBytes(of: componentHeader,
       as: RawKeyPathComponent.Header.self)
   }
+  var isPureStruct = true
+  var offset: UInt32? = nil
+      
+  for value in walker.isPureStruct {
+    isPureStruct = isPureStruct && value
+  }
+
+  // Disable the optimization in the general case of 0 components.
+  // Note that a KeyPath such as \SomeStruct.self would still technically
+  // have a valid offset of 0.
+  // TODO: Add the logic to distinguish pure struct
+  // 0-component KeyPaths (tuples too?) from others.
+  if walker.isPureStruct.count == 0 {
+    isPureStruct = false
+  }
```

Comment:
A zero-component key path should be the `\.self` keypath, which could be treated as a "pure struct" key path with offset of zero. Did that happen to cause problems in practice?

Reply:
From what I remember, it was when you started off with an identity KeyPath to a reference type, and then tried to append a pure-struct value-type KeyPath to it. 
I'll take another quick look at this.